hamilflow/models/harmonic_oscillator_chain.py

```diff
@@ -0,0 +1,142 @@
+    $Y_k = Y^*_{-k \mod N}$, see [Wikipedia](https://en.wikipedia.org/wiki/Discrete_Fourier_transform#DFT_of_real_and_purely_imaginary_signals).
+    """
+
+    def __init__(
+        self,
+        omega: float,
+        initial_conditions: Sequence[Mapping[str, float | tuple[float, float]]],
+        odd_dof: bool,
+    ) -> None:
+        """Instantiate an oscillator chain.
+
+        :param omega: frequence parameter
+        :param initial_conditions: a sequence of initial conditions on the Fourier modes.
+            The first element in the sequence is that of the zero mode, taking a position and a velocity.
+            Rest of the elements are that of the independent travelling waves, taking two amplitudes and two initial phases.
+        :param odd_dof: The system will have `2 * len(initial_conditions) + int(odd_dof) - 2` degrees of freedom.
+        """
+        self.n_dof = 2 * len(initial_conditions) + odd_dof - 2
+        if not odd_dof:
+            prefix = "For even degrees of freedom, "
+            if self.n_dof == 0:
+                raise ValueError(prefix + "at least 1 travelling wave is needed")
+            amp = cast(tuple[float, float], initial_conditions[-1]["amp"])
+            if amp[0] != amp[1]:
+                msg = "k == N // 2 must have equal positive and negative amplitudes."
+                raise ValueError(prefix + msg)
+        self.omega = omega
+        self.odd_dof = odd_dof
+
+        self.free_mode = FreeParticle(cast(Mapping[str, float], initial_conditions[0]))
+
+        self.independent_csho_modes = [
+            self._sho_factory(
+                k,
+                cast(tuple[float, float], ic["amp"]),
+                cast(tuple[float, float] | None, ic.get("phi")),
+            )
+            for k, ic in enumerate(initial_conditions[1:], 1)
+        ]
+
+    def _sho_factory(
+        self,
+        k: int,
+        amp: tuple[float, float],
+        phi: tuple[float, float] | None = None,
+    ) -> ComplexSimpleHarmonicOscillator:
+        return ComplexSimpleHarmonicOscillator(
+            dict(
+                omega=2 * self.omega * np.sin(np.pi * k / self.n_dof),
+            ),
+            dict(x0=amp) | (dict(phi=phi) if phi else {}),
+        )
+
+    @cached_property
+    def definition(
+        self,
+    ) -> dict[
+        str,
+        float
+        | int
+        | dict[str, dict[str, float | list[float]]]
+        | list[dict[str, dict[str, float | tuple[float, float]]]],
+    ]:
+        """model params and initial conditions defined as a dictionary."""
+        return dict(
+            omega=self.omega,
+            n_dof=self.n_dof,
+            free_mode=self.free_mode.definition,
+            independent_csho_modes=[
+                rwm.definition for rwm in self.independent_csho_modes
+            ],
+        )
+
+    def _z(
+        self, t: "Sequence[float] | ArrayLike[float]"
```

Comment:
I just noticed this. Is there a reason that we use quotes here? Was there a parsing error?

Reply:
py310 can easily get unhappy without the quotation marks for no obvious reason. Is there an argument against using quotation marks?

Reply:
> py310 can easily get unhappy without the quotation marks for no obvious reason. Is there an argument against using quotation marks?

not sure, just thinking that the hint becomes a string and not sure if mypy or IDE works in the way it intended to.

I know people annotate returned self using str but not sure what happens for args.

Reply:
> > py310 can easily get unhappy without the quotation marks for no obvious reason. Is there an argument against using quotation marks?
> 
> not sure, just thinking that the hint becomes a string and not sure if mypy or IDE works in the way it intended to.
> 
> I know people annotate returned self using str but not sure what happens for args. 

- vscode works fine with the hint being a string. 
- it seems quoted hints belong to so-called forward reference https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING

Reply:
Then it is good.